src/psyclone/psyir/backend/sir.py

```diff
@@ -0,0 +1,408 @@
+
+
+def gen_stencil(node):
+    '''Given an array access as input, determine the form of stencil
+    access and return it in the form expected by the SIR as a
+    string. Raise an exception if the array access is not a recognised
+    stencil access.
+
+    :param node: an array access.
+    :type node: :py:class:`psyclone.psyGen.Array`
+
+    :returns: the SIR stencil access format for the array access.
+    :rtype: str
+
+    :raises VisitorError: if the node is not the expected type or the \
+    array access is not in a recognised stencil form.
+
+    '''
+    if not isinstance(node, Array):
+        raise VisitorError(
+            "gen_stencil expected an Array as input but found '{0}'."
+            "".format(type(node)))
+    dims = []
+    for child in node.children:
+        if isinstance(child, Reference):
+            dims.append("0")
+        elif isinstance(child, BinaryOperation):
+            if isinstance(child.children[0], Reference) and \
+               isinstance(child.children[1], Literal):
+                if child.operator.name == "SUB":
+                    dims.append("-"+child.children[1].value)
+                elif child.operator.name == "ADD":
+                    dims.append(child.children[1].value)
+                else:
+                    raise VisitorError(
+                        "gen_stencil unsupported stencil operator found "
+                        "'{0}'. Expecting '+' or '-'."
+                        "".format(child.operator.name))
+            else:
+                raise VisitorError(
+                    "gen_stencil unsupported stencil index found '{0}'."
+                    "".format(str(child)))
+        else:
+            raise VisitorError(
+                "gen_stencil unsupported (non-stencil) index found '{0}'."
+                "".format(str(child)))
+    return "[{0}]".format(",".join(dims))
+
+
+class SIRWriter(PSyIRVisitor):
+    '''Implements a PSyIR-to-SIR back end for PSyIR kernel code (not
+    currently PSyIR algorithm code which has its own gen method for
+    generating Fortran).
+
+    :param bool skip_nodes: if skip_nodes is False then an exception \
+    is raised if a visitor method for a PSyIR node has not been \
+    implemented, otherwise the visitor continues, printing out a \
+    representation of the unsupported node. This is an optional \
+    argument which defaults to False.
+    :param indent_string: specifies what to use for indentation. This \
+    is an optional argument that defaults to two spaces.
+    :type indent_string: str or NoneType
+    :param int initial_indent_depth: Specifies how much indentation to \
+    start with. This is an optional argument that defaults to 0.
+
+    '''
+    def __init__(self, skip_nodes=False, indent_string="  ",
+                 initial_indent_depth=0):
+        super(SIRWriter, self).__init__(skip_nodes, indent_string,
+                                        initial_indent_depth)
+        self._field_names = set()
```

Comment:
Please add a comment describing what `self._field_names` is for.

Reply:
Done.